pkg/api/db.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		log.Error("Get volume failed in create volume attachment method: ", err)
+		return nil, err
+	}
+	if vol.Status != model.VOLUME_AVAILABLE {
+		errMsg := "Only the status of volume is available, attachment can be created"
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+	if in.Id == "" {
+		in.Id = uuid.NewV4().String()
+	}
+	if in.CreatedAt == "" {
+		in.CreatedAt = time.Now().Format(constants.TimeFormat)
+	}
+	if len(in.AdditionalProperties) == 0 {
+		in.AdditionalProperties = map[string]interface{}{"attachment": "attachment"}
+	}
+	if len(in.ConnectionData) == 0 {
+		in.ConnectionData = map[string]interface{}{"attachment": "attachment"}
+	}
+	if in.Platform == "" {
+		in.Platform = ""
+	}
+	if in.OsType == "" {
+		in.OsType = ""
+	}
+	if in.Ip == "" {
+		in.Ip = ""
+	}
+	if in.Host == "" {
+		in.Host = ""
+	}
+	if in.Initiator == "" {
+		in.Initiator = ""
+	}
+
+	var atc = &model.VolumeAttachmentSpec{
+		BaseModel: &model.BaseModel{
+			Id:        in.Id,
+			CreatedAt: in.CreatedAt,
+		},
+		VolumeId: in.VolumeId,
+		HostInfo: model.HostInfo{
+			Platform:  in.Platform,
+			OsType:    in.OsType,
+			Ip:        in.Ip,
+			Host:      in.Host,
+			Initiator: in.Initiator,
+		},
+		Status:         model.VOLUMEATM_CREATING,
+		Metadata:       utils.MergeStringMaps(in.Metadata, vol.Metadata),
+		ConnectionInfo: in.ConnectionInfo,
+	}
+
+	result, err := db.C.CreateVolumeAttachment(ctx, atc)
+	if err != nil {
+		log.Error("Error occurred in dock module when create volume attachment in db:", err)
+		return nil, err
+	}
+	return result, nil
+}
+
+func CreateVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) (*model.VolumeSnapshotSpec, error) {
+	vol, err := db.C.GetVolume(ctx, in.VolumeId)
+	if err != nil {
+		log.Error("Get volume failed in create volume snapshot method: ", err)
+		return nil, err
+	}
+	if vol.Status != model.VOLUME_AVAILABLE && vol.Status != model.VOLUME_IN_USE {
+		var errMsg = "Only the status of volume is available or in-use, the snapshot can be created"
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+
+	if in.Id == "" {
+		in.Id = uuid.NewV4().String()
+	}
+
+	if in.CreatedAt == "" {
+		in.CreatedAt = time.Now().Format(constants.TimeFormat)
+	}
+
+	var snap = &model.VolumeSnapshotSpec{
+		BaseModel: &model.BaseModel{
+			Id:        in.Id,
+			CreatedAt: in.CreatedAt,
+		},
+		Name:        in.Name,
+		Description: in.Description,
+		VolumeId:    in.VolumeId,
+		Size:        vol.Size,
+		Metadata:    utils.MergeStringMaps(in.Metadata, vol.Metadata),
+		Status:      model.VOLUMESNAP_CREATING,
+	}
+
+	result, err := db.C.CreateVolumeSnapshot(ctx, snap)
+	if err != nil {
+		log.Error("Error occurred in dock module when create volume snapshot in db:", err)
+		return nil, err
+	}
+	return result, nil
+}
+
+func DeleteVolumeSnapshotDBEntry(ctx *c.Context, in *model.VolumeSnapshotSpec) error {
+	if in.Status != model.VOLUMESNAP_AVAILABLE {
```

Comment:
The logic is not right, you need to find the real volume snapshot in database by calling `GetVolumeSnapshot` method first. You can't never find the status of volume snapshot from `in`.

Reply:
Please see the code yourself.